tracer/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp

```diff
@@ -600,14 +599,41 @@ HRESULT STDMETHODCALLTYPE CorProfiler::ModuleLoadFinished(ModuleID module_id, HR
             ModuleMetadata(metadata_import, metadata_emit, assembly_import, assembly_emit, module_info.assembly.name,
                            module_info.assembly.app_domain_id, &corAssemblyProperty);
 
-        Logger::Info("ModuleLoadFinished: ", managed_profiler_name, " - Fix PInvoke maps");
+        const auto& assemblyImport = GetAssemblyImportMetadata(assembly_import);
+        const auto& assemblyVersion = assemblyImport.version.str();
+
+        Logger::Info("ModuleLoadFinished: ", managed_profiler_name, " v", assemblyVersion, " - Fix PInvoke maps");
 #ifdef _WIN32
         RewritingPInvokeMaps(module_metadata, windows_nativemethods_type);
         RewritingPInvokeMaps(module_metadata, appsec_windows_nativemethods_type);
 #else
         RewritingPInvokeMaps(module_metadata, nonwindows_nativemethods_type);
         RewritingPInvokeMaps(module_metadata, appsec_nonwindows_nativemethods_type);
 #endif // _WIN32
+
+        if (IsVersionCompatibilityEnabled())
+        {
+            // No need to rewrite if the target assembly matches the expected version
+            if (assemblyImport.version != managed_profiler_assembly_reference->version)
+            {
+                // Don't rewrite on .NET Core if the target assembly has a higher version than the expected
+                if (!runtime_information_.is_core() ||
+                    assemblyImport.version < managed_profiler_assembly_reference->version)
```

Comment:
[nit] I think I'd find this expression easier to grok if it was inverted, especially as the comment describes the inverted situation. Alternatively, invert the comment 😉

Reply:
Done in https://github.com/DataDog/dd-trace-dotnet/pull/2020/commits/e1b26a25a9095a2a389a61a409e792c1b84a20b4